urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/UpperclassHousingSelection.java

```diff
@@ -0,0 +1,246 @@
+            theInstance=new UpperclassHousingSelection();
+        }
+        return theInstance;
+    }
+
+
+    /** 
+     * Assign the set of students (presumably non-incoming-freshmen, but
+     * this is not checked) passed to their dorm rooms, in a way that does 
+     * not take race into account. (compare {@link #assignByRace}.) As a 
+     * precursor (side effect) to this, any students already existing in 
+     * upperclass dorms will be removed.
+     */
+    public void assign(Bag students){
+
+        emptyDorms();
+
+        // Purge all freshmen. (We're not assigning those.)
+        Bag upperclassmen = null;
+        try { 
+            upperclassmen = (Bag) students.clone(); 
+        } catch (Exception e) { 
+            e.printStackTrace(); 
+            System.exit(1); 
+        }
+        for(int x=upperclassmen.size()-1; x>=0; x--){
+            Student s = (Student) upperclassmen.get(x);
+            if (s.getGrade() < 2) {
+                upperclassmen.remove(s);
+            }
+        }
+        
+        for(int x=0;x<upperclassmen.size();x++){
+            Student s = (Student) upperclassmen.get(x);
+            if (s.getGrade() < 2) {
+                // We're only assigning upperclassmen here.
+                continue;
+            }
+            s.leaveRoom();
+            for(int y=0; y<dorms.size();y++){
+                if(s.hasRoom()){break;}
+                Dorm d = (Dorm) dorms.get(y);
+                if(d.isFull()){
+                    continue;
+                }
+                if(d.isFemaleOnly() && s.getGender()==Student.Gender.MALE){
+                    continue;
+                }
+                int i = 0;
+                while(i<d.getNumRooms()){
+                    if(s.hasRoom()){break;}
+                    try{
+                        d.getRoomByIndex(i).addResident(s);
+                    }catch(IllegalArgumentException e){//System.out.println(e);
+                    }
+                    i++;
+                }
+            }
+        }
+
+        //Error testing to see if there are upperclassmen who don't have a 
+        //  room/dorms that aren't full
+        // int q=0;
+        for(int x=0;x<upperclassmen.size();x++){
+            Student s = (Student) upperclassmen.get(x);
+            if(!s.hasRoom()){
+                System.out.println(s + " has no upperclass dorm room!");
+            }
+        }
+    }
+
+    /** 
+     * Assign the set of students (presumably non-incoming-freshmen, but 
+     * this is not checked) students passed to their dorm rooms, in a way 
+     * that <i>does</i> take race into account. (compare {@link #assign}.)
+     * Minorities will have a higher probability of rooming with other
+     * minorities than they would otherwise have had. As a precursor (side 
+     * effect) to this, any students already existing in upperclass dorms
+     * will be removed.
+     */
+    public void assignByRace(Bag upperclassmen) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.